### PR TITLE
Mitigate AddProvider verse AddEntityFrameworkProvider method confusion

### PR DESCRIPTION
**Pull request by ajcvickers**
By hiding the infrastructure method from Intellisense and adding docs.

Also adds Cosmos and In-Memory sugar.

Part of #24743

### CODE REVIEW

File: src/EFCore.InMemory/Extensions/InMemoryServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -24,12 +26,69 @@ namespace Microsoft.Extensions.DependencyInjection
...
+        /// <typeparam name="TContext"> The type of context to be registered. </typeparam>
+        /// <param name="serviceCollection"> The <see cref="IServiceCollection" /> to add services to. </param>
+        /// <param name="databaseName">
+        ///     The name of the in-memory database. This allows the scope of the in-memory database to be controlled
+        ///     independently of the context. The in-memory database is shared anywhere the same name is used.
+        /// </param>
+        /// <param name="inMemoryOptionsAction"> An optional action to allow additional in-memory database specific configuration. </param>
+        /// <param name="optionsAction"> An optional action to configure the <see cref="DbContextOptions" /> for the context. </param>
+        /// <returns> The same service collection so that multiple calls can be chained. </returns>
+        public static IServiceCollection AddInMemoryDatabase<TContext>(
```
Comment (DamianEdwards) [inline src/EFCore.InMemory/Extensions/InMemoryServiceCollectionExtensions.cs:64]: oooo I like how this reads 😃

**Comment by roji** [comment]
We could also have the analyzer flag usages of AddEntityFrameworkProvider - opened #25455 to track this.

File: src/EFCore.Cosmos/Extensions/CosmosServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -24,12 +26,68 @@ namespace Microsoft.Extensions.DependencyInjection
     /// </summary>
     public static class CosmosServiceCollectionExtensions
     {
+        /// <summary>
+        ///     <para>
+        ///         Registers the given Entity Framework <see cref="DbContext"/> as a service in the <see cref="IServiceCollection" />
```
Comment (roji) [inline src/EFCore.Cosmos/Extensions/CosmosServiceCollectionExtensions.cs:31]: Here and below:

```suggestion
        ///         Registers the given Entity Framework <see cref="DbContext" /> as a service in the <see cref="IServiceCollection" />
```

**Review by roji — APPROVED** [review]

File: src/EFCore.Cosmos/Extensions/CosmosServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -24,12 +26,68 @@ namespace Microsoft.Extensions.DependencyInjection
...
+        ///         and configures it to to connect to a Azure Cosmos database.
```
Comment (stevendarby) [inline src/EFCore.Cosmos/Extensions/CosmosServiceCollectionExtensions.cs:32]: Doubled up “to” and should be “an Azure”

**Review by AndriySvyryd — APPROVED** [review]

File: src/EFCore.Cosmos/Extensions/CosmosServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -24,12 +26,68 @@ namespace Microsoft.Extensions.DependencyInjection
...
+        }
+        
         /// <summary>
         ///     <para>
         ///         Adds the services required by the Azure Cosmos database provider for Entity Framework
         ///         to an <see cref="IServiceCollection" />.
         ///     </para>
         ///     <para>
+        ///         Warning: Do not call this method accidentally. It is much more likely you need
+        ///         to call <see cref="AddCosmos{TContext}"/>.
```
Comment (roji) [inline src/EFCore.Cosmos/Extensions/CosmosServiceCollectionExtensions.cs:88]:
```suggestion
        ///         to call <see cref="AddCosmos{TContext}" />.
```

**Comment by ajcvickers** [comment]
Note from triage: remove `AddInMemoryDatabase` because we want to [direct people away from the in-memory database](https://docs.microsoft.com/en-us/ef/core/testing/).

**Comment by DamianEdwards** [comment]
@ajcvickers that's a shame as the in-memory database can be a good candidate during tutorials when the complexity of initializing and managing a storage provider adds more complexity/concepts to the learning stage than is desired. In-memory being self-initializing and requiring no separate acquisition or setup is appealing in these scenarios.

@davidfowl @LadyNaggaga

**Comment by davidfowl** [comment]
> Note from triage: remove AddInMemoryDatabase because we want to direct people away from the in-memory database.

Why would we remove this? It's so simple to get started and avoids the need for migrations until required.

**Comment by AndriySvyryd** [comment]
> Why would we remove this? It's so simple to get started and avoids the need for migrations until required.

Because it's not a real database. A lot of users expect it to have relational semantics or being production-ready. InMemoryDatabase has a very limited scope and we should steer them towards SqlServer LocalDB or Sqlite in-memory.

Microsoft has been criticized for building toy apps in tutorials using techniques that aren't acceptable in production-ready code, InMemoryDatabase certainly falls into this category.

**Comment by davidfowl** [comment]
It's already in another package you have to explicitly reference. Not adding this method doesn't do anything. I think we should add the method. 

You can continue to push people to SQLite but don't neuter the package arbitrarily

**Comment by ajcvickers** [comment]
Split out the in-memory question to #25487 so that the rest can be merged. We can continue discussion there.

**Pull request merged**